File: go/ch29/improvement_log.go
// go/ch29/improvement_log.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"path/filepath"
)

type DayEntry struct {
	Day     int    `json:"day"`
	Date    string `json:"date"`
	Owner   string `json:"owner"`
	Score   int    `json:"score"`
	Notes   string `json:"notes"`
	Commits int    `json:"commits"`
}

type WeeklyLog struct {
	Contest string     `json:"contest"`
	Problem string     `json:"problem"`
	Team    string     `json:"team"`
	Unit    string     `json:"unit"`
	Days    []DayEntry `json:"days"`
}

func loadLog(path string) (WeeklyLog, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return WeeklyLog{}, err
	}
	var log WeeklyLog
	if err := json.Unmarshal(data, &log); err != nil {
		return WeeklyLog{}, err
	}
	return log, nil
}

func summarize(log WeeklyLog) {
	if len(log.Days) == 0 {
		fmt.Fprintln(os.Stderr, "no days in log")
		return
	}
	scores := make([]int, len(log.Days))
	for i, d := range log.Days {
		scores[i] = d.Score
	}
	best := scores[0]
	bestDay := log.Days[0].Day
	for i, s := range scores {
		if s > best {
			best = s
			bestDay = log.Days[i].Day
		}
	}
	first, last := scores[0], scores[len(scores)-1]
	delta := last - first
	pct := 0.0
	if first != 0 {
		pct = 100.0 * float64(delta) / float64(first)
	}

	fmt.Printf("contest: %s\n", log.Contest)
	fmt.Printf("problem: %s  team: %s\n", log.Problem, log.Team)
	fmt.Printf("days: %d  unit: %s\n\n", len(log.Days), log.Unit)
	fmt.Println("day  date        owner    score     delta    note")

	prev := -1
	for _, d := range log.Days {
		deltaS := ""
		if prev >= 0 {
			deltaS = fmt.Sprintf("%+d", d.Score-prev)
		}
		note := d.Notes
		if len(note) > 28 {
			note = note[:28]
		}
		fmt.Printf("%3d  %-10s  %-8s  %9d  %7s  %s\n",
			d.Day, d.Date, d.Owner, d.Score, deltaS, note)
		prev = d.Score
	}
	fmt.Println()
	fmt.Printf("start → end: %d → %d  (Δ %+d, %+.2f%%)\n", first, last, delta, pct)
	fmt.Printf("best: %d on day %d\n", best, bestDay)

	regressions := 0
	for i := 1; i < len(scores); i++ {
		if scores[i] < scores[i-1] {
			regressions++
		}
	}
	fmt.Printf("regression days: %d\n", regressions)
	commits := 0
	for _, d := range log.Days {
		commits += d.Commits
	}
	fmt.Fprintf(os.Stderr, "total commits: %d\n", commits)
}

func main() {
	input := flag.String("input", filepath.Join("data", "ch29", "weekly_log.json"), "weekly log JSON")
	flag.Parse()

	log, err := loadLog(*input)
	if err != nil {
		fmt.Fprintf(os.Stderr, "read error: %v\n", err)
		os.Exit(1)
	}
	summarize(log)
}

File: python/ch29/improvement_log.py
# python/ch29/improvement_log.py
"""週次改善ログ JSON を読み、スコア推移を要約（第29章）."""
from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any, Dict, List


def default_log_path() -> Path:
    return Path(__file__).resolve().parents[2] / "data" / "ch29" / "weekly_log.json"


def load_log(path: Path) -> Dict[str, Any]:
    with path.open(encoding="utf-8") as f:
        return json.load(f)


def summarize(log: Dict[str, Any]) -> None:
    days: List[Dict[str, Any]] = log.get("days", [])
    if not days:
        print("no days in log", file=sys.stderr)
        return

    scores = [int(d["score"]) for d in days]
    best = max(scores)
    best_day = days[scores.index(best)]["day"]
    first, last = scores[0], scores[-1]
    delta = last - first
    pct = (100.0 * delta / first) if first else 0.0

    print(f"contest: {log.get('contest', '?')}")
    print(f"problem: {log.get('problem', '?')}  team: {log.get('team', '?')}")
    print(f"days: {len(days)}  unit: {log.get('unit', 'score')}")
    print()
    print("day  date        owner    score     delta    note")
    prev = None
    for d in days:
        sc = int(d["score"])
        delta_s = "" if prev is None else f"{sc - prev:+d}"
        note = str(d.get("notes", ""))[:28]
        print(
            f"{d.get('day', '?'):>3}  {d.get('date', '?'):10}  "
            f"{d.get('owner', '?'):8}  {sc:>9}  {delta_s:>7}  {note}"
        )
        prev = sc
    print()
    print(f"start → end: {first} → {last}  (Δ {delta:+d}, {pct:+.2f}%)")
    print(f"best: {best} on day {best_day}")
    regressions = sum(
        1 for i in range(1, len(scores)) if scores[i] < scores[i - 1]
    )
    print(f"regression days: {regressions}")
    total_commits = sum(int(d.get("commits", 0)) for d in days)
    print(f"total commits: {total_commits}", file=sys.stderr)


def main() -> None:
    path = Path(sys.argv[1]) if len(sys.argv) > 1 else default_log_path()
    log = load_log(path)
    summarize(log)


if __name__ == "__main__":
    main()
